### PR TITLE
Benchmarks - Add distributed inference benchmark

### PR DESCRIPTION
**Pull request by yzygitzh**
**Description**
This PR adds a micro-benchmark of distributed model inference workloads.

**Major Revision**
- Add a new micro-benchmark dist-inference.
- Add corresponding example and unit tests.
- Update configuration files to include this new micro-benchmark.
- Update micro-benchmark README.

### CODE REVIEW

File: superbench/config/azure/inference/standard_nc96ads_a100_v4.yaml [hunk 1]
```diff
@@ -101,6 +101,8 @@ superbench:
     computation-communication-overlap:
       <<: *default_pytorch_mode
       enable: false
+    dist-inference-simulation:
```
Comment (cp5555) [inline superbench/config/azure/inference/standard_nc96ads_a100_v4.yaml:104]: dist-inference

Comment (cp5555) [reply superbench/config/azure/inference/standard_nc96ads_a100_v4.yaml:104]: Please remove all simulation

File: superbench/config/azure_ndmv4.yaml [hunk 2]
```diff
@@ -159,6 +159,8 @@ superbench:
       <<: *default_pytorch_mode
     computation-communication-overlap:
       <<: *default_pytorch_mode
+    dist-inference-simulation:
```
Comment (cp5555) [inline superbench/config/azure_ndmv4.yaml:162]: remove `simulation`

File: superbench/config/azure/inference/standard_nc96ads_a100_v4.yaml (hunk 1 above)
Comment (cp5555) [reply superbench/config/azure/inference/standard_nc96ads_a100_v4.yaml:104]: Also, do we need to add NDv4 and NDmv4 only?

File: docs/user-tutorial/benchmarks/micro-benchmarks.md [hunk 3]
```diff
@@ -394,6 +394,23 @@ Test the performance of large scale matmul operation with multiple GPUs:
 | pytorch-sharding-matmul/allreduce_time | time (ms) | Time of sharding matmul using allreduce. |
 | pytorch-sharding-matmul/allgather_time | time (ms) | Time of sharding matmul using allgather. |
 
+####
+
+### `dist-inference-simulation`
+
+#### Introduction
+
+Test the performance of a simulated distributed model inference.
```
Comment (cp5555) [inline docs/user-tutorial/benchmarks/micro-benchmarks.md:403]: Test the performance of distributed model inference.

File: superbench/benchmarks/micro_benchmarks/__init__.py [hunk 4]
```diff
@@ -57,4 +58,5 @@
     'ShardingMatmul',
     'TCPConnectivityBenchmark',
     'TensorRTInferenceBenchmark',
+    'DistInferenceSimulation'
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/__init__.py:61]: DistInference

File: superbench/config/azure/inference/standard_nc96ads_a100_v4.yaml (hunk 1 above)
Comment (yzygitzh) [reply superbench/config/azure/inference/standard_nc96ads_a100_v4.yaml]: Fixed

File: superbench/config/azure_ndmv4.yaml (hunk 2 above)
Comment (yzygitzh) [reply superbench/config/azure_ndmv4.yaml]: Fixed

File: superbench/benchmarks/micro_benchmarks/__init__.py (hunk 4 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/__init__.py]: Fixed

File: docs/user-tutorial/benchmarks/micro-benchmarks.md (hunk 3 above)
Comment (yzygitzh) [reply docs/user-tutorial/benchmarks/micro-benchmarks.md]: Fixed

File: superbench/benchmarks/micro_benchmarks/dist_inference.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        test_step_times = []
+        for i in range(self._args.num_steps):
+            start = self.__timer()
+            model.forward(data)
+            end = self.__timer()
+            test_step_times.append((end - start) * 1000)
+
+        if not self._process_numeric_result('warmup_step_times', warmup_step_times, cal_percentile=True):
+            return False
+        if not self._process_numeric_result('test_step_times', test_step_times, cal_percentile=True):
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/dist_inference.py:335]: We should have each layer metrics as well.

Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference.py]: Fixed

File: superbench/benchmarks/micro_benchmarks/__init__.py [hunk 6]
```diff
@@ -29,32 +29,13 @@
...
-    'RocmMemBwBenchmark',
-    'ShardingMatmul',
-    'TCPConnectivityBenchmark',
-    'TensorRTInferenceBenchmark',
+    'ComputationCommunicationOverlap', 'CpuMemBwLatencyBenchmark', 'CpuStreamBenchmark', 'CublasBenchmark',
+    'CublasLtBenchmark', 'CudaGemmFlopsBenchmark', 'CudaMemBwBenchmark', 'CudaNcclBwBenchmark', 'CudnnBenchmark',
+    'DiskBenchmark', 'GPCNetBenchmark', 'GemmFlopsBenchmark', 'GpuBurnBenchmark', 'GpuCopyBwBenchmark', 'IBBenchmark',
+    'IBLoopbackBenchmark', 'KernelLaunch', 'MemBwBenchmark', 'MicroBenchmark', 'MicroBenchmarkWithInvoke',
+    'ORTInferenceBenchmark', 'RocmGemmFlopsBenchmark', 'RocmMemBwBenchmark', 'ShardingMatmul',
+    'TCPConnectivityBenchmark', 'TensorRTInferenceBenchmark', 'DistInference'
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/__init__.py:40]: Why change it?

Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/__init__.py:40]: Need to pass code lint

File: superbench/benchmarks/micro_benchmarks/__init__.py [hunk 7]
```diff
@@ -30,33 +30,13 @@
...
-    'RocmMemBwBenchmark',
-    'ShardingMatmul',
-    'TCPConnectivityBenchmark',
-    'TensorRTInferenceBenchmark',
+    'ComputationCommunicationOverlap', 'CpuMemBwLatencyBenchmark', 'CpuHplBenchmark', 'CpuStreamBenchmark',
+    'CublasBenchmark', 'CublasLtBenchmark', 'CudaGemmFlopsBenchmark', 'CudaMemBwBenchmark', 'CudaNcclBwBenchmark',
+    'CudnnBenchmark', 'DiskBenchmark', 'GPCNetBenchmark', 'GemmFlopsBenchmark', 'GpuBurnBenchmark',
+    'GpuCopyBwBenchmark', 'IBBenchmark', 'IBLoopbackBenchmark', 'KernelLaunch', 'MemBwBenchmark', 'MicroBenchmark',
+    'MicroBenchmarkWithInvoke', 'ORTInferenceBenchmark', 'RocmGemmFlopsBenchmark', 'RocmMemBwBenchmark',
+    'ShardingMatmul', 'TCPConnectivityBenchmark', 'TensorRTInferenceBenchmark', 'DistInference'
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/__init__.py:41]: Do we need to recover this?

File: superbench/benchmarks/micro_benchmarks/dist_inference.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+            '--hidden_size',
+            type=int,
+            default=1024,
+            required=False,
+            help='Hidden size.',
+        )
+        self._parser.add_argument(
+            '--num_layers',
+            type=int,
+            default=10,
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/dist_inference.py:187]: Maybe 1 for quick test

File: superbench/benchmarks/micro_benchmarks/dist_inference.py [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+            required=False,
+            help='Number of warmup steps.',
+        )
+        self._parser.add_argument(
+            '--num_steps',
+            type=int,
+            default=200,
+            required=False,
+            help='Number of test steps.',
+        )
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/dist_inference.py:232]: Maybe set warm_up and num_step as a practical number

**Review by cp5555 — APPROVED** [review]

File: superbench/benchmarks/micro_benchmarks/__init__.py (hunk 7 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/__init__.py:41]: Lint check will fail if recover

Comment (cp5555) [reply superbench/benchmarks/micro_benchmarks/__init__.py:41]: Why it works previously?

Comment (abuccts) [reply superbench/benchmarks/micro_benchmarks/__init__.py:41]: need "," in the end, and pls sort it alphabetically (after diskbenchmark)

Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/__init__.py]: Fixed

File: superbench/benchmarks/micro_benchmarks/dist_inference.py (hunk 8 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference.py]: Fixed

File: superbench/benchmarks/micro_benchmarks/dist_inference.py (hunk 9 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference.py:232]: Fixed

File: docs/user-tutorial/benchmarks/micro-benchmarks.md [hunk 10]
```diff
@@ -409,6 +409,21 @@ Test the performance of large scale matmul operation with multiple GPUs:
 | pytorch-sharding-matmul/allreduce_time | time (ms) | Time of sharding matmul using allreduce. |
 | pytorch-sharding-matmul/allgather_time | time (ms) | Time of sharding matmul using allgather. |
 
+####
```
Comment (abuccts) [inline docs/user-tutorial/benchmarks/micro-benchmarks.md:412]: remove this line?

File: superbench/benchmarks/micro_benchmarks/dist_inference.py [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+        self.bias = torch.rand(self.hidden_size, dtype=getattr(torch, precision.value), device=device)
+        self.num_ranks = num_ranks
+        self.step_times = []
+
+        self.__init_computation_kernels(computation)
+        self.__init_communication_kernels(communication)
+        self.__init_activation_kernels(activation)
+
+    def __init_computation_kernels(self, computation):
+        """Select computation kernel according to option."""
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/dist_inference.py:74]: docstring for args?

File: docs/user-tutorial/benchmarks/micro-benchmarks.md [hunk 12]
```diff
@@ -409,6 +409,21 @@ Test the performance of large scale matmul operation with multiple GPUs:
...
+#### Introduction
+
+Test the performance of distributed model inference.
+
+#### Metrics
+
+| Name                                                 | Unit      | Description                                           |
+|------------------------------------------------------|-----------|-------------------------------------------------------|
+| pytorch-dist-inference/test_step_times               | time (ms) | Average time of test inference steps.                 |
+| pytorch-dist-inference/test_step_times_${percentile} | time (ms) | Tail (50,90,95,99,99.9) time of test inference steps. |
```
Comment (abuccts) [inline docs/user-tutorial/benchmarks/micro-benchmarks.md:425]: remove the prefix "test_" in metrics?

File: docs/user-tutorial/benchmarks/micro-benchmarks.md (hunk 10 above)
Comment (yzygitzh) [reply docs/user-tutorial/benchmarks/micro-benchmarks.md]: Fixed.

File: superbench/benchmarks/micro_benchmarks/dist_inference.py (hunk 11 above)
Comment (yzygitzh) [reply superbench/benchmarks/micro_benchmarks/dist_inference.py]: Fixed.

File: docs/user-tutorial/benchmarks/micro-benchmarks.md (hunk 12 above)
Comment (yzygitzh) [reply docs/user-tutorial/benchmarks/micro-benchmarks.md]: Fixed.

**Review by abuccts — APPROVED** [review]

**Pull request merged**